Sources/Views/MessageLabel.swift

```diff
@@ -23,6 +23,298 @@
  */
 
 import UIKit
+extension NSTextStorage {
+    
+    var attributedString: NSAttributedString {
+        let range = NSRange(location: 0, length: mutableString.length)
+        return attributedSubstring(from: range)
+    }
+    
+    var mutableAttributedString: NSMutableAttributedString {
+        return NSMutableAttributedString(attributedString: attributedString)
+    }
+    
+    func addAttribute(key: NSAttributedStringKey, value: AnyObject, range: NSRange? = nil) {
+        let range = range ?? NSRange(location: 0, length: length)
+        addAttribute(key, value: value, range: range)
+    }
+    
+}
+
+open class AttributeLabel: UIView {
+    
+    // MARK: - Text Internals
+    
+    private var layoutManager: NSLayoutManager
+    
+    private var textContainer: NSTextContainer
+    
+    private var textStorage: NSTextStorage
+    
+    // MARK: - Detected Ranges
+    
+    private var phoneRanges: [NSRange] = []
+    
+    private var addressRanges: [NSRange] = []
+    
+    private var dateRanges: [NSRange] = []
+    
+    private var urlRanges: [NSRange] = []
+    
+    // MARK: - Properties [Public]
+    
+    public var attributedText: NSAttributedString? {
+        get {
+            return textStorage.attributedString
+        }
+        set {
+            let string = newValue ?? NSAttributedString(string: "")
+            textStorage.setAttributedString(string)
+            
+            setStorageByParsing(text: string) {
+                self.setNeedsDisplay()
+            }
+            
+            setNeedsDisplay()
+        }
+    }
+    
+    public var text: String? {
+        get {
+            return textStorage.string
+        }
+        set {
+            let string = addLocalAttributes(to: newValue)
+            textStorage.setAttributedString(string)
+            
+            setStorageByParsing(text: string) {
+                self.setNeedsDisplay()
+            }
+            
+            setNeedsDisplay()
+        }
+    }
+    
+    public var font: UIFont = UIFont.systemFont(ofSize: 10.0) {
+        didSet {
+            textStorage.addAttribute(key: .font, value: font)
+            // Doesn't apply to detected ranges
+            addAttributesToTextStorage(for: enabledDetectors)
```

Comment:
probably overkill to apply all the attributes again. instead we should only apply the font attribute that changed